File: internal/functions/hmac_test.go
package functions

import (
	"errors"
	"net/http"
	"strings"
	"testing"
	"time"
)

// Closes layer 3 of advisory GHSA-7428-mvpp-rhr7. These tests verify
// the Go signer / verifier pair is internally consistent. The
// cross-language compatibility (Go signs, Deno verifies — and vice
// versa) is asserted by a fixed reference vector at the bottom; if
// either side's canonical-message format drifts, the vector breaks.

const testSecret = "0123456789abcdef0123456789abcdef" // 32 bytes

func makeTestHeaders() http.Header {
	h := http.Header{}
	h.Set("X-Project-ID", "p-001")
	h.Set("X-Schema-Name", "tenant_abc")
	h.Set("X-Function-ID", "fn-deadbeef")
	h.Set("X-User-ID", "u-1234")
	h.Set("X-User-Email", "alice@example.com")
	h.Set("X-Plan", "pro")
	h.Set("X-Request-ID", "req-xyz")
	return h
}

func TestNewSigner_RejectsShortSecret(t *testing.T) {
	for _, short := range []string{"", "abc", strings.Repeat("a", 31)} {
		if _, err := NewSigner(short); err == nil {
			t.Errorf("NewSigner(%q) expected error, got nil", short)
		}
	}
}

func TestNewSigner_AcceptsAdequateSecret(t *testing.T) {
	if _, err := NewSigner(testSecret); err != nil {
		t.Errorf("NewSigner(32-byte) failed: %v", err)
	}
}

func TestSign_AddsTimestampAndSignatureHeaders(t *testing.T) {
	signer, _ := NewSigner(testSecret)
	h := makeTestHeaders()
	signer.Sign(h, time.Unix(1700000000, 0))

	if h.Get("X-Eurobase-Timestamp") != "1700000000" {
		t.Errorf("expected timestamp 1700000000, got %q", h.Get("X-Eurobase-Timestamp"))
	}
	sig := h.Get("X-Eurobase-Signature")
	if len(sig) != 64 {
		t.Errorf("expected 64-hex-char signature, got %d chars: %q", len(sig), sig)
	}
}

func TestSignVerify_RoundTrip(t *testing.T) {
	signer, _ := NewSigner(testSecret)
	h := makeTestHeaders()
	signer.Sign(h, time.Now())

	if err := signer.Verify(h, VerifyOptions{}); err != nil {
		t.Errorf("Verify of self-signed headers failed: %v", err)
	}
}

func TestVerify_RejectsTamperedHeader(t *testing.T) {
	signer, _ := NewSigner(testSecret)
	h := makeTestHeaders()
	signer.Sign(h, time.Now())

	// Flip the schema name. Verifier should detect.
	h.Set("X-Schema-Name", "tenant_other")
	if err := signer.Verify(h, VerifyOptions{}); !errors.Is(err, ErrSignatureMismatch) {
		t.Errorf("Verify of tampered headers should return ErrSignatureMismatch, got %v", err)
	}
}

func TestVerify_RejectsMissingSignature(t *testing.T) {
	signer, _ := NewSigner(testSecret)
	h := makeTestHeaders()
	// No Sign() call — headers have no signature.
	err := signer.Verify(h, VerifyOptions{})
	if !errors.Is(err, ErrMissingSignature) {
		t.Errorf("Verify of unsigned headers should return ErrMissingSignature, got %v", err)
	}
}

func TestVerify_RejectsExpiredTimestamp(t *testing.T) {
	signer, _ := NewSigner(testSecret)
	h := makeTestHeaders()
	signer.Sign(h, time.Unix(1700000000, 0)) // signature is valid for that ts

	// "Now" is well outside the 5-min skew window.
	err := signer.Verify(h, VerifyOptions{
		MaxClockSkew: 5 * time.Minute,
		Now:          func() time.Time { return time.Unix(1700001000, 0) }, // +1000s
	})
	if !errors.Is(err, ErrTimestampOutOfWindow) {
		t.Errorf("Verify of expired timestamp should return ErrTimestampOutOfWindow, got %v", err)
	}
}

func TestVerify_RejectsFutureTimestamp(t *testing.T) {
	signer, _ := NewSigner(testSecret)
	h := makeTestHeaders()
	signer.Sign(h, time.Unix(1700001000, 0))

	err := signer.Verify(h, VerifyOptions{
		MaxClockSkew: 5 * time.Minute,
		Now:          func() time.Time { return time.Unix(1700000000, 0) }, // -1000s
	})
	if !errors.Is(err, ErrTimestampOutOfWindow) {
		t.Errorf("Verify of future timestamp should return ErrTimestampOutOfWindow, got %v", err)
	}
}

func TestVerify_RejectsBadTimestampFormat(t *testing.T) {
	signer, _ := NewSigner(testSecret)
	h := makeTestHeaders()
	signer.Sign(h, time.Now())
	h.Set("X-Eurobase-Timestamp", "not-a-number")
	if err := signer.Verify(h, VerifyOptions{}); !errors.Is(err, ErrTimestampOutOfWindow) {
		t.Errorf("expected ErrTimestampOutOfWindow on bad timestamp, got %v", err)
	}
}

func TestVerify_AcceptsWithinSkewWindow(t *testing.T) {
	signer, _ := NewSigner(testSecret)
	h := makeTestHeaders()
	signer.Sign(h, time.Unix(1700000000, 0))

	// 4 minutes 30 seconds later — within 5-min window.
	err := signer.Verify(h, VerifyOptions{
		MaxClockSkew: 5 * time.Minute,
		Now:          func() time.Time { return time.Unix(1700000270, 0) },
	})
	if err != nil {
		t.Errorf("Verify within skew window should succeed, got %v", err)
	}
}

func TestVerify_DifferentSecretFails(t *testing.T) {
	signer1, _ := NewSigner(testSecret)
	signer2, _ := NewSigner("ffffffffffffffffffffffffffffffff") // 32 chars but different key
	h := makeTestHeaders()
	signer1.Sign(h, time.Now())
	if err := signer2.Verify(h, VerifyOptions{}); !errors.Is(err, ErrSignatureMismatch) {
		t.Errorf("Verify with different secret should return ErrSignatureMismatch, got %v", err)
	}
}

func TestSign_OrderOfHeadersDoesNotMatter(t *testing.T) {
	// http.Header iteration is map-ordered; canonicalMessage uses
	// fixed lookup keys so the map iteration order can't affect the
	// output. Two signers seeded with the same secret should produce
	// the same signature for the same logical input regardless of
	// which order the test sets the keys.
	signer, _ := NewSigner(testSecret)
	now := time.Unix(1700000000, 0)

	a := http.Header{}
	a.Set("X-User-Email", "alice@example.com")
	a.Set("X-Project-ID", "p-001")
	a.Set("X-Schema-Name", "tenant_abc")
	a.Set("X-Function-ID", "fn-deadbeef")
	a.Set("X-User-ID", "u-1234")
	a.Set("X-Plan", "pro")
	a.Set("X-Request-ID", "req-xyz")
	signer.Sign(a, now)

	b := makeTestHeaders()
	signer.Sign(b, now)

	if a.Get("X-Eurobase-Signature") != b.Get("X-Eurobase-Signature") {
		t.Errorf("signatures should match regardless of header set order; got %q vs %q",
			a.Get("X-Eurobase-Signature"), b.Get("X-Eurobase-Signature"))
	}
}

// TestCanonicalMessage_ReferenceVector is the cross-language compat
// pin. If this changes, the runner's `functions-runner/hmac.ts` MUST
// change in lockstep — otherwise gateway-signed requests fail to verify.
func TestCanonicalMessage_ReferenceVector(t *testing.T) {
	h := makeTestHeaders()
	got := canonicalMessage(h, "1700000000")
	want := "v=1\n" +
		"ts=1700000000\n" +
		"project=p-001\n" +
		"schema=tenant_abc\n" +
		"function=fn-deadbeef\n" +
		"user=u-1234\n" +
		"email=alice@example.com\n" +
		"plan=pro\n" +
		"requestid=req-xyz"
	if got != want {
		t.Errorf("canonicalMessage drift detected.\n got:\n%q\n want:\n%q", got, want)
	}
}

func TestCanonicalMessage_EmptyValuesEmittedAsEmpty(t *testing.T) {
	// X-User-ID / X-User-Email are blank for unauthenticated requests.
	// They MUST be present in the canonical message as empty strings,
	// not omitted — otherwise a forged request that adds an arbitrary
	// X-User-ID could mutate the canonical form to match a different
	// signature.
	h := http.Header{}
	h.Set("X-Project-ID", "p")
	h.Set("X-Schema-Name", "t")
	h.Set("X-Function-ID", "f")
	h.Set("X-Plan", "free")
	h.Set("X-Request-ID", "r")
	got := canonicalMessage(h, "100")
	if !strings.Contains(got, "user=\n") {
		t.Errorf("expected user= line with empty value, got:\n%s", got)
	}
	if !strings.Contains(got, "email=\n") {
		t.Errorf("expected email= line with empty value, got:\n%s", got)
	}
}

File: functions-runner/hmac_test.ts
// Closes GHSA-7428-mvpp-rhr7 layer 3 — runner-side HMAC verification.
//
// Run locally:
//   deno test --no-check hmac_test.ts
//
// Cross-language compat: the canonical-message format and the HMAC
// algorithm match `internal/functions/hmac.go` byte-for-byte. The
// reference vector at the bottom of this file uses the same expected
// signature as `TestCanonicalMessage_ReferenceVector` in the Go suite —
// if the Go side ever drifts, this test fails too, surfacing the
// drift early.

import { assert, assertEquals, assertStrictEquals } from "https://deno.land/std@0.224.0/assert/mod.ts";
import { canonicalMessage, constantTimeEqualHex, newVerifier } from "./hmac.ts";

// 32-byte secret (matches the Go test).
const SECRET = "0123456789abcdef0123456789abcdef";

function makeHeaders(overrides: Record<string, string> = {}): Headers {
  const h = new Headers();
  h.set("X-Project-ID", "p-001");
  h.set("X-Schema-Name", "tenant_abc");
  h.set("X-Function-ID", "fn-deadbeef");
  h.set("X-User-ID", "u-1234");
  h.set("X-User-Email", "alice@example.com");
  h.set("X-Plan", "pro");
  h.set("X-Request-ID", "req-xyz");
  for (const [k, v] of Object.entries(overrides)) h.set(k, v);
  return h;
}

// signMatchesGo computes an HMAC for a given canonical message + secret
// the same way the Go signer does, so we can construct test fixtures
// without spinning up a Go process.
async function signMatchesGo(secret: string, msg: string): Promise<string> {
  const enc = new TextEncoder();
  const key = await crypto.subtle.importKey(
    "raw",
    enc.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const macBytes = new Uint8Array(await crypto.subtle.sign("HMAC", key, enc.encode(msg)));
  let hex = "";
  for (const b of macBytes) hex += b.toString(16).padStart(2, "0");
  return hex;
}

Deno.test("newVerifier rejects short secrets", async () => {
  for (const short of ["", "abc", "a".repeat(31)]) {
    let threw = false;
    try {
      await newVerifier(short);
    } catch {
      threw = true;
    }
    assert(threw, `newVerifier(${JSON.stringify(short)}) should throw`);
  }
});

Deno.test("newVerifier accepts a 32-byte secret", async () => {
  await newVerifier(SECRET); // shouldn't throw
});

Deno.test("verify accepts a correctly-signed request", async () => {
  const verifier = await newVerifier(SECRET);
  const ts = "1700000000";
  const headers = makeHeaders();
  headers.set("X-Eurobase-Timestamp", ts);
  const sig = await signMatchesGo(SECRET, canonicalMessage(headers, ts));
  headers.set("X-Eurobase-Signature", sig);
  const result = await verifier.verify(headers, {
    now: () => 1700000000,
    maxSkewSeconds: 60,
  });
  assertStrictEquals(result.ok, true);
});

Deno.test("verify rejects tampered headers", async () => {
  const verifier = await newVerifier(SECRET);
  const ts = "1700000000";
  const headers = makeHeaders();
  headers.set("X-Eurobase-Timestamp", ts);
  const sig = await signMatchesGo(SECRET, canonicalMessage(headers, ts));
  headers.set("X-Eurobase-Signature", sig);
  // Flip schema name.
  headers.set("X-Schema-Name", "tenant_other");
  const result = await verifier.verify(headers, { now: () => 1700000000 });
  assertStrictEquals(result.ok, false);
  if (!result.ok) assertEquals(result.reason, "mismatch");
});

Deno.test("verify rejects missing signature headers", async () => {
  const verifier = await newVerifier(SECRET);
  const headers = makeHeaders();
  const result = await verifier.verify(headers);
  assertStrictEquals(result.ok, false);
  if (!result.ok) assertEquals(result.reason, "missing");
});

Deno.test("verify rejects out-of-window timestamp", async () => {
  const verifier = await newVerifier(SECRET);
  const ts = "1700000000";
  const headers = makeHeaders();
  headers.set("X-Eurobase-Timestamp", ts);
  const sig = await signMatchesGo(SECRET, canonicalMessage(headers, ts));
  headers.set("X-Eurobase-Signature", sig);
  // 1000 seconds ahead of timestamp — outside default 300s window.
  const result = await verifier.verify(headers, { now: () => 1700001000 });
  assertStrictEquals(result.ok, false);
  if (!result.ok) assertEquals(result.reason, "out_of_window");
});

Deno.test("verify rejects bad timestamp format", async () => {
  const verifier = await newVerifier(SECRET);
  const headers = makeHeaders();
  headers.set("X-Eurobase-Timestamp", "not-a-number");
  headers.set("X-Eurobase-Signature", "0".repeat(64));
  const result = await verifier.verify(headers);
  assertStrictEquals(result.ok, false);
  if (!result.ok) assertEquals(result.reason, "bad_timestamp");
});

Deno.test("verify rejects a signature signed with a different secret", async () => {
  const verifier = await newVerifier(SECRET);
  const otherSecret = "ffffffffffffffffffffffffffffffff"; // 32 chars but different
  const ts = "1700000000";
  const headers = makeHeaders();
  headers.set("X-Eurobase-Timestamp", ts);
  const sig = await signMatchesGo(otherSecret, canonicalMessage(headers, ts));
  headers.set("X-Eurobase-Signature", sig);
  const result = await verifier.verify(headers, { now: () => 1700000000 });
  assertStrictEquals(result.ok, false);
  if (!result.ok) assertEquals(result.reason, "mismatch");
});

Deno.test("constantTimeEqualHex returns true for equal strings", () => {
  assertStrictEquals(constantTimeEqualHex("abc123", "abc123"), true);
});

Deno.test("constantTimeEqualHex returns false for unequal strings", () => {
  assertStrictEquals(constantTimeEqualHex("abc123", "abc124"), false);
  assertStrictEquals(constantTimeEqualHex("abc", "abcdef"), false);
});

Deno.test("canonicalMessage matches the Go reference vector", () => {
  const headers = makeHeaders();
  const got = canonicalMessage(headers, "1700000000");
  const want = "v=1\n" +
    "ts=1700000000\n" +
    "project=p-001\n" +
    "schema=tenant_abc\n" +
    "function=fn-deadbeef\n" +
    "user=u-1234\n" +
    "email=alice@example.com\n" +
    "plan=pro\n" +
    "requestid=req-xyz";
  assertEquals(got, want);
});

Deno.test("canonicalMessage emits empty string for missing fields", () => {
  // Mirrors the Go test — the line MUST be present so a forger can't
  // mutate the canonical form by inserting a new value into a previously
  // empty field.
  const h = new Headers();
  h.set("X-Project-ID", "p");
  h.set("X-Schema-Name", "t");
  h.set("X-Function-ID", "f");
  h.set("X-Plan", "free");
  h.set("X-Request-ID", "r");
  const msg = canonicalMessage(h, "100");
  assert(msg.includes("user=\n"), "missing empty user= line");
  assert(msg.includes("email=\n"), "missing empty email= line");
});
